### PR TITLE
[build_base][RLlib] PPO TF RL Module with Policy

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: avnish <avnish@anyscale.com>

Implementation of the ppo tf rl policy with rl module and tests for performance / compilation

https://tensorboard.dev/experiment/CnF7klfnTvyQ7UqQ2e7XlA/#scalars&tagFilter=reward&_smoothingWeight=0

here is the tensorboard comparing the old implementation with the new implementation.

The new implementation is not compatible with tf eager tracing at this moment. I suspect that its related to the issue of sample batches having their count zeroed out by eager tracing. UPDATE: It is in fact this issue, but there is an issue with using tf conditionals to execute either forward_exploration or forward inference inside of eager_tf_policy_v2. forward_exploration and forward_inference would both need to return the same type. 

This can all be resolved if we only tf.function wrap the individual modules instead of wrapping the whole policy class, which we'll be able to do when we rewrite the sampler and stop using the policy class all together.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        del dist_class
+        fwd_out = model.forward_train(train_batch)
+
+        curr_action_dist = fwd_out[SampleBatch.ACTION_DIST]
+        dist_class = curr_action_dist.__class__
+        value_fn_out = fwd_out[SampleBatch.VF_PREDS]
+
+        reduce_mean_valid = tf.reduce_mean
+
+        prev_action_dist = dist_class(
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:93]: These distribution classes are not consistent with torch version. The idea was for distribution classes to be explicit about what they need, to get constructed, similar to how torch is. For example if I want a categorial distribution, it's either `Categorical(logits=...)` or `Categorical(probs=...)`. We should have ideally fixed this in the PPOTfRLModulePR but here it is again. When we use the new distribution classes ActionDistributionInputs should be a non-flat nested dict instead of a flat tensor that you have assumed here. While what you have works totally fine, it introduces inconsistencies between this and the torch version. It's your call. We can move on with this until we get both torch and tf working and then consolidate the learnings and make them consistent at the very end, or make them consistent now. Either is fine with me.

Comment (kouroshHakha) [reply rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:93]: As you can see, we have made totally different choices about how we want to construct our model output distributions. This is the power of RLModule. It doesn't bound you to any assumptions about the model output/input formats. Despite its flexibility it also has a mechanism for checking the input/output structures.

Comment (kouroshHakha) [reply rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:93]: If distribution class here is a type of class that actually uses model as part of its logic, this line of code would break, because you are now passing an RLModule in instead of a model.

File: rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py [hunk 2]
```diff
@@ -60,6 +60,19 @@ def tearDownClass(cls):
     def test_ppo_compilation_and_schedule_mixins(self):
         """Test whether PPO can be built with all frameworks."""
 
+        model_config = {
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:63]: IKR

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 3]
```diff
@@ -243,18 +243,18 @@ def input_specs_exploration(self):
 
     @override(RLModule)
     def output_specs_exploration(self) -> SpecDict:
-        specs = {SampleBatch.ACTION_DIST: self.__get_action_dist_type()}
         if self._is_discrete:
-            specs[SampleBatch.ACTION_DIST_INPUTS] = {
-                "logits": TorchTensorSpec("b, h", h=int(self.config.action_space.n))
-            }
+            action_dist_inputs = ((SampleBatch.ACTION_DIST_INPUTS, "logits"),)
```
Comment (kouroshHakha) [inline rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:247]: ok much much nicer :)

File: rllib/algorithms/ppo/ppo.py [hunk 4]
```diff
@@ -124,6 +124,10 @@ def get_default_rl_module_class(self) -> Union[Type["RLModule"], str]:
             )
 
             return PPOTorchRLModule
+        elif self.framework_str == "tf2":
+            from ray.rllib.algorithms.ppo.tf.ppo_tf_rl_module import PPOTfModule
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/ppo.py:128]: fixed

File: rllib/evaluation/postprocessing.py [hunk 5]
```diff
@@ -205,7 +205,7 @@ def compute_gae_for_sample_batch(
             # will not be needed but takes time to compute.
             input_dict = policy._lazy_tensor_dict(input_dict)
             fwd_out = policy.model.forward_exploration(input_dict)
-            last_r = fwd_out[SampleBatch.VF_PREDS]
+            last_r = fwd_out[SampleBatch.VF_PREDS][0]
```
Comment (kouroshHakha) [inline rllib/evaluation/postprocessing.py:208]: wait, does this work for both torch and tf? If you are making this change here, what is the change on the other side to make torch not break?

File: rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py [hunk 6]
```diff
@@ -79,7 +92,7 @@ def test_ppo_compilation_and_schedule_mixins(self):
                 ),
             )
             .rollouts(
-                num_rollout_workers=1,
+                num_rollout_workers=0,
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:95]: reverting

File: rllib/utils/tf_utils.py [hunk 7]
```diff
@@ -570,7 +570,7 @@ def zero_logps_from_actions(actions: TensorStructType) -> TensorType:
 
 @DeveloperAPI
 def warn_if_infinite_kl_divergence(
-    policy: Type["TFPolicy"], mean_kl_loss: TensorType
+    policy: Type["TFPolicy"], mean_kl: TensorType
```
Comment (kouroshHakha) [inline rllib/utils/tf_utils.py:573]: curious! why renaming?

File: rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py [hunk 8]
```diff
@@ -60,6 +60,19 @@ def tearDownClass(cls):
     def test_ppo_compilation_and_schedule_mixins(self):
         """Test whether PPO can be built with all frameworks."""
 
+        model_config = {
+            "torch": dict(
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:64]: I don't support all of the model config kwargs that the PPO torch rl module does since I don't have support for lstms.

I also don't have defaults for things like vf_share_layers, activations, and number of hiddens. I want these to be configured by the user explicitly.

[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:64]: I mean for torch only. They are supported, at least on the RLModule side of things. lstm is false in the test, but all we had to do once fully supported was to add lstm=True to the test. Now the config does not match what we originally had on master.

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py (hunk 1 above)
Comment (avnishn) [reply rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:93]: yeah this right here is a hack so that I would stay unblocked without having to add a new distributions class for now.

There is one distributions class that I can think of that would use the model. That is the auto regressive distribution.

File: rllib/evaluation/postprocessing.py (hunk 5 above)
Comment (avnishn) [reply rllib/evaluation/postprocessing.py:208]: oh you know in hindsight this should actually be fwd_out[SampleBatch.VF_PREDS][-1] in order to be correct. 

last r is supposed to be the value estimation of the last observation.

I don't know why this didn't break the torch one, but the torch tests seem to be passing.

File: rllib/utils/tf_utils.py (hunk 7 above)
Comment (avnishn) [reply rllib/utils/tf_utils.py:573]: the variable that gets passed to this function is called mean_kl_loss where it is used, however mean kl loss is computed like this:

```
action_kl = prev_action_dist.kl(curr_action_dist)
mean_kl_loss = reduce_mean_valid(action_kl)
```

so mean kl loss is just actually the mean kl. And this warning function is just checking to see that the mean kl is a finite number.

Comment (avnishn) [reply rllib/utils/tf_utils.py:573]: it doesn't make sense to call it mean kl loss

File: rllib/evaluation/postprocessing.py (hunk 5 above)
Comment (avnishn) [reply rllib/evaluation/postprocessing.py]: changed to -1

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py (hunk 1 above)
Comment (avnishn) [reply rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:93]: offline conversation:

going to punt

File: rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py [hunk 9]
```diff
@@ -148,7 +149,7 @@ def test_ppo_exploration_setup(self):
         obs = np.array(0)
 
         # TODO (Kourosh) Test against all frameworks.
-        for fw in framework_iterator(config, frameworks=("torch")):
+        for fw in framework_iterator(config, frameworks=("torch", "tf2")):
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:152]: oh actually in this case it doesn't matter because I'm using the framework iterator in kinda a dumb way here. I'm not using it to modify an algo config, I'm just using it to get a framework name.

setting eager_tracing to false only matters if we were modifying an algo config.

File: rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py [hunk 10]
```diff
@@ -90,33 +103,21 @@ def test_ppo_compilation_and_schedule_mixins(self):
 
         num_iterations = 2
 
-        # TODO (Kourosh): for now just do torch
         for fw in framework_iterator(
-            config, frameworks=("torch"), with_eager_tracing=True
+            config, frameworks=("torch", "tf2"), with_eager_tracing=False
```
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:107]: sure

[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/algorithms/ppo/tests/test_ppo_with_rl_module.py:118]: done

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        train_batch: SampleBatch,
+    ) -> Union[TensorType, List[TensorType]]:
+        del dist_class
+        fwd_out = model.forward_train(train_batch)
+
+        curr_action_dist = fwd_out[SampleBatch.ACTION_DIST]
+        dist_class = curr_action_dist.__class__
+        value_fn_out = fwd_out[SampleBatch.VF_PREDS]
+
+        reduce_mean_valid = tf.reduce_mean
```
Comment (gjoliver) [inline rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py:91]: what's the point of saving this in a local variable?

**Review by gjoliver — APPROVED** [review]

File: rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py (hunk 11 above)
Comment (avnishn) [reply rllib/algorithms/ppo/tf/ppo_tf_policy_rlm.py]: I got rid of this because of the redundancy

**Review by kouroshHakha — APPROVED** [review]
